### ISSUE TITLE
types: Header.Hash returns nil hash if there ValidatorsHash is not set with security implication perhaps

### ISSUE THREAD

**Issue by odeke-em** [tendermint/tendermint#1302]
I was working on tests for lite and security hardening and noticed that Header.Hash returns nil if its ValidatorsHash field is not set
https://github.com/tendermint/tendermint/blob/c8a2bdf78ba7aaaf4284fa78c1b9b05c5e7342bc/types/block.go#L179-L181

The intentions of that code might be related to the protocol and expectations but if lite is invoking Hash to compare blocks directly yet it returns the same hash for blocks might have different attributes, IMHO this is a security flaw and for all we know a nefarious person who knows that we use lite is using `tendermint lite` and sending over blocks without a ValidatorsHash.

TL;DR: an external auditor who doesn't know anything about Tendermint should be able to be comfortable verifying that entirely different blocks should NOT be validated and equal

```go
package main

import (
	"testing"
	"time"

	"github.com/tendermint/tendermint/lite"
	"github.com/tendermint/tendermint/lite/proxy"
	"github.com/tendermint/tendermint/types"
)

func TestLiteValidationIsFallacious(t *testing.T) {
	block := &types.Block{
		Header: &types.Header{
			Height: 11,
			Time:   time.Date(2018, 1, 1, 1, 1, 1, 1, time.UTC),
		},
		Data: new(types.Data),
	}
	commit := lite.Commit{
		Header: &types.Header{
			Height: 11,
			Time:   time.Date(2017, 1, 1, 1, 1, 1, 1, time.UTC),
		},
	}

	if err := proxy.ValidateBlock(block, commit); err == nil {
		t.Errorf("Pwned!")
	}
}
```
Will print `Pwned!` if we don't change the underlying code

And in full
```go
package main

import (
	"bytes"
	"testing"
	"time"

	"github.com/tendermint/tendermint/lite"
	"github.com/tendermint/tendermint/lite/proxy"
	"github.com/tendermint/tendermint/types"
)

func TestLiteValidationIsFallacious(t *testing.T) {
	block := &types.Block{
		Header: &types.Header{
			Height: 11,
			Time:   time.Date(2018, 1, 1, 1, 1, 1, 1, time.UTC),
		},
		Data: new(types.Data),
	}
	commit := lite.Commit{
		Header: &types.Header{
			Height: 11,
			Time:   time.Date(2017, 1, 1, 1, 1, 1, 1, time.UTC),
		},
	}

	if err := proxy.ValidateBlock(block, commit); err == nil {
		t.Errorf("Pwned!")
	}
}

func TestHeaderFails(t *testing.T) {
	tests := []struct {
		hdr         *types.Header
		wantNilHash bool
	}{
		{hdr: nil, wantNilHash: true},
		{hdr: &types.Header{}, wantNilHash: true},
		{
			hdr: &types.Header{
				Height: 100,
				Time:   time.Now().UTC(),
			},
			wantNilHash: false,
		},
		{
			hdr: &types.Header{
				Height: 100,
				Time:   time.Now().UTC(),
			},
			wantNilHash: false,
		},
		{
			hdr: &types.Header{
				Height: 111,
				Time:   time.Date(2018, 12, 2, 1, 1, 3, 1, time.UTC),
			},
			wantNilHash: false,
		},
	}

	for i, tt := range tests {
		hdr := tt.hdr
		if hdr == nil {
			hdr = new(types.Header)
		}
		hash := hdr.Hash()
		if g, w := bytes.Equal(hash, nil), tt.wantNilHash; g != w {
			t.Errorf("#%d: gotIsNilHash=%v gotHash=(% X) wantNil=%v", i, g, hash, w)
		}
	}
}
```
will print
```shell
$ go test
--- FAIL: TestLiteValidationIsFallacious (0.00s)
	tf_test.go:29: Pwned!
--- FAIL: TestHeaderFails (0.00s)
	tf_test.go:70: #2: gotIsNilHash=true gotHash=() wantNil=false
	tf_test.go:70: #3: gotIsNilHash=true gotHash=() wantNil=false
	tf_test.go:70: #4: gotIsNilHash=true gotHash=() wantNil=false
FAIL
exit status 1
FAIL	_/Users/emmanuelodeke/Desktop/experiments/tendermint-hardening/header-hash	0.014s
```

**Comment by ebuchman** [comment]
Yep, the lite client should check for a nil Header.Hash and throw an error. Thanks!

**Comment by xla** [comment]
@jaekwon Still relevant?

**Comment by ebuchman** [comment]
> TL;DR: an external auditor who doesn't know anything about Tendermint should be able to be comfortable verifying that entirely different blocks should NOT be validated and equal

I think this is reasonable. We should probably change this behaviour and fix the lite package to deal with it.

While we're at it maybe we can make SignedHeader not take a pointer, in the spirit of https://github.com/tendermint/tendermint/issues/1346

**Comment by melekes** [comment]
no longer relevant, I believe.
